File: executor/signal_scorer.py
"""Signal scoring and grading — port from TypeScript lib/signal-scorer.ts."""

from __future__ import annotations

from constants import (
    SCORE_A_GRADE_A_MIN, SCORE_A_GRADE_B_MIN, SCORE_A_GRADE_C_MIN,
    SCORE_B_GRADE_A_MIN, SCORE_B_GRADE_B_MIN, SCORE_B_GRADE_C_MIN,
    SCORE_C_GRADE_A_MIN, SCORE_C_GRADE_B_MIN, SCORE_C_GRADE_C_MIN,
    STRATEGY_B_LOW_VOL_THRESHOLD, STRATEGY_B_LOW_VOL_PENALTY,
    STRATEGY_C_BELOW_KUMO_PENALTY,
)


def _to_grade(score: int, a_min: int, b_min: int, c_min: int) -> str:
    if score >= a_min:
        return "A"
    if score >= b_min:
        return "B"
    if score >= c_min:
        return "C"
    return "D"


def score_strategy_a(result, ctx: dict) -> tuple[int, str]:
    """Score Strategy A (N-Day High Breakout) result. Returns (score, grade)."""
    if not result.triggered:
        return 0, "D"

    score = 45  # base

    # Breakout strength: moderate breakouts outperform extreme ones
    if result.breakout_pct is not None:
        if 0.5 <= result.breakout_pct <= 2.5:
            score += 15  # sweet spot — not too early, not overextended
        elif result.breakout_pct < 0.5:
            score += 5   # marginal breakout
        else:
            score += 8   # overextended — higher risk of reversal

    # Volume surge: moderate surge is optimal
    if result.volume_ratio is not None:
        if 1.3 <= result.volume_ratio <= 2.0:
            score += 15  # healthy confirmation volume
        elif result.volume_ratio > 2.0:
            score += 8   # extreme volume can signal exhaustion
        else:
            score += 5

    adx = ctx.get("adx")
    if adx is not None and adx >= 25:
        score += 10

    regime = ctx.get("market_regime", "neutral")
    if regime == "bull":
        score += 10
    elif regime == "bear":
        score -= 20

    if ctx.get("stoch_cross_up"):
        score += 5

    score = max(0, min(100, score))
    return score, _to_grade(score, SCORE_A_GRADE_A_MIN, SCORE_A_GRADE_B_MIN, SCORE_A_GRADE_C_MIN)


def score_strategy_b(result, ctx: dict) -> tuple[int, str]:
    """Score Strategy B (Deep Oversold Reversal) result. Returns (score, grade)."""
    if not result.triggered:
        return 0, "D"

    score = 40  # base

    # RSI depth bonus: deeper oversold = stronger signal
    if result.rsi is not None:
        if result.rsi <= 25:
            score += 25
        elif result.rsi <= 28:
            score += 20
        else:
            score += 15

    # Strong reversal candle is core — already required for trigger
    if result.strong_reversal:
        score += 5

    # Both confirmations present = highest quality
    if result.volume_spike and result.macd_improving:
        score += 15
    elif result.macd_improving:
        score += 10
    elif result.volume_spike:
        score += 8

    regime = ctx.get("market_regime", "neutral")
    if regime == "bull":
        score += 5
    elif regime == "bear":
        score -= 15

    if ctx.get("stoch_cross_up"):
        score += 8

    score = max(0, min(100, score))
    return score, _to_grade(score, SCORE_B_GRADE_A_MIN, SCORE_B_GRADE_B_MIN, SCORE_B_GRADE_C_MIN)


def score_strategy_c(result, ctx: dict) -> tuple[int, str]:
    """Score Strategy C result. Returns (score, grade)."""
    if not result.triggered:
        return 0, "D"

    score = 45  # base

    if result.gc_days_ago == 1:
        score += 15
    elif result.gc_days_ago == 2:
        score += 10
    elif result.gc_days_ago == 3:
        score += 7
    else:
        score += 4

    adx = ctx.get("adx")
    if adx is not None and adx >= 25:
        score += 10

    regime = ctx.get("market_regime", "neutral")
    if regime == "bull":
        score += 10
    elif regime == "bear":
        score -= 20

    if ctx.get("price_above_kumo"):
        score += 5
    if ctx.get("price_below_kumo"):
        score += STRATEGY_C_BELOW_KUMO_PENALTY

    if ctx.get("sar_bull"):
        score += 3

    score = max(0, min(100, score))
    return score, _to_grade(score, SCORE_C_GRADE_A_MIN, SCORE_C_GRADE_B_MIN, SCORE_C_GRADE_C_MIN)

File: lib/signal-scorer.ts
import type { StrategyAResult } from '@/lib/strategies/strategy-a';

type MarketRegime = 'bull' | 'bear' | 'neutral';
import type { StrategyBResult } from '@/lib/strategies/strategy-b';
import type { StrategyCResult } from '@/lib/strategies/strategy-c';
import {
  SCORE_A_GRADE_A_MIN, SCORE_A_GRADE_B_MIN, SCORE_A_GRADE_C_MIN,
  SCORE_B_GRADE_A_MIN, SCORE_B_GRADE_B_MIN, SCORE_B_GRADE_C_MIN,
  SCORE_C_GRADE_A_MIN, SCORE_C_GRADE_B_MIN, SCORE_C_GRADE_C_MIN,
  STRATEGY_B_LOW_VOL_THRESHOLD, STRATEGY_B_LOW_VOL_PENALTY,
  STRATEGY_C_BELOW_KUMO_PENALTY,
} from '@/lib/constants';
import type { SignalGrade } from '@/lib/types/database';

type ScoringContext = {
  adx: number | null;
  stochCrossUp: boolean;
  priceAboveKumo?: boolean;
  priceBelowKumo?: boolean;
  sarBull?: boolean;
  volumeRatio: number | null;
  rsi: number | null;
  macdHistogram: number | null;
  marketRegime: MarketRegime;
};

export type SignalScore = {
  score: number;
  grade: SignalGrade;
  breakdown: Record<string, number>;
};

export function scoreStrategyA(
  result: StrategyAResult,
  ctx: ScoringContext,
): SignalScore {
  const breakdown: Record<string, number> = {};
  let score = 0;

  if (!result.triggered) {
    return { score: 0, grade: 'D', breakdown: { base: 0 } };
  }

  breakdown.base = 50; score += 50;

  if (result.components.rsiCrossDaysAgo === 1) { breakdown.rsi_cross = 15; score += 15; }
  else if (result.components.rsiCrossDaysAgo === 2) { breakdown.rsi_cross = 10; score += 10; }

  if (result.components.volumeSurge) { breakdown.volume = 10; score += 10; }
  if (ctx.adx !== null && ctx.adx >= 25) { breakdown.adx = 10; score += 10; }
  if (ctx.marketRegime === 'bull') { breakdown.market = 10; score += 10; }
  if (ctx.marketRegime === 'bear') { breakdown.market = -20; score -= 20; }
  if (ctx.stochCrossUp) { breakdown.stoch = 5; score += 5; }

  score = Math.max(0, Math.min(100, score));
  return { score, grade: toGrade(score, SCORE_A_GRADE_A_MIN, SCORE_A_GRADE_B_MIN, SCORE_A_GRADE_C_MIN), breakdown };
}

export function scoreStrategyB(
  result: StrategyBResult,
  ctx: ScoringContext,
): SignalScore {
  const breakdown: Record<string, number> = {};
  let score = 0;

  if (!result.triggered) {
    return { score: 0, grade: 'D', breakdown: { base: 0 } };
  }

  breakdown.base = 55; score += 55;

  if (ctx.rsi !== null && ctx.rsi <= 25) { breakdown.rsi = 20; score += 20; }
  else if (ctx.rsi !== null && ctx.rsi <= 30) { breakdown.rsi = 15; score += 15; }
  else if (ctx.rsi !== null && ctx.rsi <= 35) { breakdown.rsi = 10; score += 10; }
  else if (ctx.rsi !== null && ctx.rsi <= 38) { breakdown.rsi = 5; score += 5; }

  if (result.components.macdImproving) { breakdown.macd = 10; score += 10; }
  if (ctx.volumeRatio !== null && ctx.volumeRatio >= 1.5) { breakdown.volume = 10; score += 10; }
  if (ctx.volumeRatio !== null && ctx.volumeRatio < STRATEGY_B_LOW_VOL_THRESHOLD) {
    breakdown.low_volume = STRATEGY_B_LOW_VOL_PENALTY; score += STRATEGY_B_LOW_VOL_PENALTY;
  }

  if (ctx.marketRegime === 'bull') { breakdown.market = 5; score += 5; }
  if (ctx.marketRegime === 'bear') { breakdown.market = -10; score -= 10; }
  if (ctx.stochCrossUp) { breakdown.stoch = 5; score += 5; }

  score = Math.max(0, Math.min(100, score));
  return { score, grade: toGrade(score, SCORE_B_GRADE_A_MIN, SCORE_B_GRADE_B_MIN, SCORE_B_GRADE_C_MIN), breakdown };
}

export function scoreStrategyC(
  result: StrategyCResult,
  ctx: ScoringContext,
): SignalScore {
  const breakdown: Record<string, number> = {};
  let score = 0;

  if (!result.triggered) {
    return { score: 0, grade: 'D', breakdown: { base: 0 } };
  }

  breakdown.base = 45; score += 45;

  if (result.components.gcDaysAgo === 1) { breakdown.gc_recency = 15; score += 15; }
  else if (result.components.gcDaysAgo === 2) { breakdown.gc_recency = 10; score += 10; }
  else if (result.components.gcDaysAgo === 3) { breakdown.gc_recency = 7; score += 7; }
  else { breakdown.gc_recency = 4; score += 4; }

  if (ctx.adx !== null && ctx.adx >= 25) { breakdown.adx = 10; score += 10; }
  if (ctx.marketRegime === 'bull') { breakdown.market = 10; score += 10; }
  if (ctx.marketRegime === 'bear') { breakdown.market = -20; score -= 20; }
  if (ctx.priceAboveKumo) { breakdown.ichimoku_kumo = 5; score += 5; }
  if (ctx.priceBelowKumo) { breakdown.ichimoku_below = STRATEGY_C_BELOW_KUMO_PENALTY; score += STRATEGY_C_BELOW_KUMO_PENALTY; }
  if (ctx.sarBull) { breakdown.sar_bull = 3; score += 3; }

  score = Math.max(0, Math.min(100, score));
  return { score, grade: toGrade(score, SCORE_C_GRADE_A_MIN, SCORE_C_GRADE_B_MIN, SCORE_C_GRADE_C_MIN), breakdown };
}

function toGrade(score: number, aMin: number, bMin: number, cMin: number): SignalGrade {
  if (score >= aMin) return 'A';
  if (score >= bMin) return 'B';
  if (score >= cMin) return 'C';
  return 'D';
}
